### PR TITLE
[LLVM] Make s_getpc_b64 rematerializable

### PR DESCRIPTION
**Pull request by Acim-Maravic**

### CODE REVIEW

**Comment by Acim-Maravic** [comment]
I found two places that can code-generate an S_GETPC_B64, one in function SIInstrInfo::insertIndirectBranch() and one in function SIInstrInfo::expandPostRAPseudo(). I think that both of these are happening after register allocation, so at that point rematerialization can not be done.

**Comment by jayfoad** [comment]
> I found two places that can code-generate an S_GETPC_B64, one in function SIInstrInfo::insertIndirectBranch() and one in function SIInstrInfo::expandPostRAPseudo(). I think that both of these are happening after register allocation, so at that point rematerialization can not be done.

Right. Rematerializing getpc is a weird case, because the rematerialized instruction will definitely _not_ have the same result as the original instruction. But I guess the only reasonable use of getpc (pre-RA) is calling the intrinsic llvm.amdgc.s.getpc and then extracting the high 32 bits of the result, which should be safe.

**Review by nhaehnle — COMMENTED** [review]
The test case doesn't actually show the `S_GETPC_B64`s as getting merged or rematerialized. I suspect we want a test structure like:
```
  %x = s_getpc_b64
  inline asm that uses approximately all SGPRs
  use %x here
```
as a pre-commit which should spill `%x` without your change. And then your change causes it to instead rematerialize the s_getpc_b64.

Keep the pre-commit in the same PR, i.e. your PR should have two commits:

1. Add the new test with the CHECK lines corresponding to behavior before your change
2. The change to isRematerializable plus changes to the test caused by it

**Comment by piotrAMD** [comment]
The option ```--stress-regalloc=2``` passed in the test has the effect of limiting the number of available registers already. 
Separating out the test as a pre-commit, as Nicolai said, will demonstrate this (hopefully!).

**Review by nhaehnle — APPROVED** [review]
Thanks for explaining. Acim, in general it would still be good to split this into two commits, but I suppose for this one it can go in as-is.

File: llvm/test/CodeGen/AMDGPU/remat-sop.mir [hunk 1]
```diff
@@ -573,3 +573,26 @@ body:             |
...
+    ; GCN-NEXT: S_NOP 0, implicit killed renamable $sgpr0_sgpr1
+    ; GCN-NEXT: S_ENDPGM 0
+    %0:sgpr_64 = S_GETPC_B64
+    %1:sgpr_64 = S_GETPC_B64
+    %2:sgpr_64 = S_GETPC_B64
+    S_NOP 0, implicit %0
+    S_NOP 0, implicit %1
+    S_NOP 0, implicit %2
+    S_ENDPGM 0
+...
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/remat-sop.mir:598]: Does this also work if it's in the bundle sequence with the following add?

Comment (jayfoad) [reply llvm/test/CodeGen/AMDGPU/remat-sop.mir:598]: You mean the SI_PC_ADD_REL_OFFSET sequence? That's not even expanded to a bundle until after RA.

**Review by nhaehnle — APPROVED** [review]

**Pull request merged**